controllers/front/ExpressCheckout.php

```diff
@@ -0,0 +1,234 @@
+use PrestaShop\Module\PrestashopCheckout\PaypalCountryCodeMatrice;
+use PrestaShop\Module\PrestashopCheckout\PaypalOrder;
+use PrestaShop\Module\PrestashopCheckout\Presenter\Cart\CartPresenter;
+
+class ps_checkoutExpressCheckoutModuleFrontController extends ModuleFrontController
+{
+    public function postProcess()
+    {
+        $isAjax = \Tools::getValue('ajax');
+
+        if ($isAjax) {
+            return false;
+        }
+
+        $token = \Tools::getValue('expressCheckoutToken');
+
+        if ($token !== \Tools::getToken()) {
+            throw new \PrestaShopException('Bad token');
+        }
+
+        $paypalOrder = \Tools::getValue('paypalOrder');
+
+        if (empty($paypalOrder)) {
+            throw new \PrestaShopException('Paypal order cannot be empty');
+        }
+
+        $paypalOrder = json_decode($paypalOrder);
+
+        if (false === $this->context->customer->isLogged()) {
+            $this->createAndLoginCustomer($paypalOrder->payer);
+        }
+
+        // Always 0 index because we are not using the paypal marketplace system
+        // This index is only used in a marketplace context
+        $this->createAddress($paypalOrder->purchase_units[0]->shipping);
+
+        $this->context->cookie->__set('paypalOrderId', $paypalOrder->id);
+        $this->context->cookie->__set('paypalEmail', $paypalOrder->payer->email_address);
+
+        $this->redirectToCheckout();
+    }
+
+    /**
+     * Handle creation and customer login
+     *
+     * @param object $payer
+     */
+    private function createAndLoginCustomer($payer)
+    {
+        $idCustomerExists = \Customer::customerExists($payer->email_address, true);
+
+        if (0 === $idCustomerExists) {
+            $customer = $this->createCustomer($payer);
+        } else {
+            $customer = new \Customer((int) $idCustomerExists);
+        }
+
+        $this->context->updateCustomer($customer);
+    }
+
+    /**
+     * Create a customer
+     *
+     * @param object $payerNode
+     *
+     * @return \Customer
+     */
+    public function createCustomer($payerNode)
+    {
+        $customer = new \Customer();
+        $customer->email = $payerNode->email_address;
+        $customer->firstname = $payerNode->name->given_name;
+        $customer->lastname = $payerNode->name->surname;
+        $customer->passwd = \Tools::passwdGen();
+        $customer->save();
+
+        return $customer;
+    }
+
+    /**
+     * Create address
+     *
+     * @param object $shipping
+     */
+    private function createAddress($shipping)
+    {
+        // check if country is available for delivery
+        $paypalIsoCode = $shipping->address->country_code;
+        $psIsoCode = (new PaypalCountryCodeMatrice())->getPrestashopIsoCode($paypalIsoCode);
+        $idCountry = \Country::getByIso($psIsoCode);
+
+        $country = new \Country($idCountry);
+
+        if ('0' == (string) $country->active) {
```

Comment:
If you take time to cast one of the variable, you can make a strict comparison with `===`.

```php
        if ('0' === (string) $country->active) {
```

But wouldn't it cleaner to check integers instead?

```php
        if (0 === (int) $country->active) {
```